kernel/memory/src/paging/mapper.rs

```diff
@@ -867,3 +867,104 @@ pub fn mapped_pages_unmap(
 pub fn mapper_from_current() -> Mapper {
     Mapper::from_current()
 }
+
+
+/// An immutably borrowed [`MappedPages`] object that derefs to `&T`.
+///
+/// When dropped, the borrow ends and the contained `MappedPages` is dropped and unmapped.
+pub struct BorrowedMappedPages<T: FromBytes> {
+    ptr: NonNull<T>,
+    mp: MappedPages,
+}
+impl<T: FromBytes> Deref for BorrowedMappedPages<T> {
+    type Target = T;
+    fn deref(&self) -> &T {
+        // SAFETY:
+        // ✅ The pointer is properly aligned because we use `NonNull`.
+        // ✅ The pointer is dereferenceable, as it has been bounds checked by `MappedPages::as_type()`.
+        // ✅ The pointer has been initialized in the constructor `try_into_borrowed()`.
+        // ✅ The lifetime of the returned reference `&T` is tied to the lifetime of the `MappedPages`,
+        //    ensuring that the `MappedPages` object will persist at least as long as the reference.
+        unsafe { self.ptr.as_ref() }
+    }
+}
+impl<T: FromBytes> BorrowedMappedPages<T> {
+    /// Immutably borrows the given `MappedPages` as an instance of type `&T` 
+    /// starting at the given `offset` into the `MappedPages`.
+    ///
+    /// See [`MappedPages::as_type()`] for more info.
+    ///
+    /// Returns an error containing the unmodified `MappedPages` and a string
+    /// describing the error.
+    pub fn try_into_borrowed(
+        mp: MappedPages,
+        offset: usize,
+    ) -> Result<BorrowedMappedPages<T>, (MappedPages, &'static str)> {
+        let mut borrowed_mp = BorrowedMappedPages {
+            ptr: NonNull::<T>::dangling(),
+            mp,
+        };
+        borrowed_mp.ptr = match borrowed_mp.mp.as_type::<T>(offset) {
+            Ok(r) => r.into(),
+            Err(e_str) => return Err((borrowed_mp.mp, e_str)),
+        };
```

Comment:
```suggestion
        let mut borrowed_mp = BorrowedMappedPages {
            ptr: match borrowed_mp.mp.as_type::<T>(offset) {
                Ok(r) => r.into(),
                Err(e_str) => return Err((borrowed_mp.mp, e_str)),
            },
            mp,
        };
```
The call to `NonNull::dangling()` seems unecessary.

Reply:
`borrowed_mp` doesn't exist yet when you use it :thinking:

Reply:
Right, good catch. I've modified the snippet to reference `mp` rather than `borrowed_mp.mp` which should work.

Reply:
I was previously moving `mp` into `borrowed_mp` in order to make it explicitly clear that the lifetime is tied to that of something owned by `borrowed_mp`, but yeah i agree it's not necessary.